File: ios/StreamHub/Core/RealtimeClient.swift
import Foundation

/// What the server pushes down `/api/realtime`.
enum RealtimeEvent: Sendable {
    case favorites(action: String)
    case progress(action: String, historyChanged: Bool)
    case receivers([CastReceiver])
    case command(from: String?, fromName: String?, command: CastCommand)
    case unknown(String)
}

/// The live connection: library sync in, remote-control commands both ways.
///
/// Reconnects on its own with exponential backoff, and knows the two close
/// codes that mean something specific — 4002 is an expired access token, which
/// has to be renewed *before* reconnecting or the client loops against the
/// server with a token it already knows is dead.
actor RealtimeClient {

    private let endpoint: URL
    private let store: SessionStore
    private let renew: @Sendable (String?) async -> Session?

    private var task: URLSessionWebSocketTask?
    private var running = false

    private(set) var sessionId: String?

    private let events: AsyncStream<RealtimeEvent>
    private let continuation: AsyncStream<RealtimeEvent>.Continuation

    private static let closeTokenExpired = 4002
    private static let closeUnauthorized = 4003

    init(baseURL: URL, store: SessionStore, renew: @escaping @Sendable (String?) async -> Session?) {
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
        components.scheme = (components.scheme == "https") ? "wss" : "ws"
        components.path = "/api/realtime"
        self.endpoint = components.url!
        self.store = store
        self.renew = renew

        var continuation: AsyncStream<RealtimeEvent>.Continuation!
        self.events = AsyncStream(bufferingPolicy: .bufferingNewest(64)) { continuation = $0 }
        self.continuation = continuation
    }

    nonisolated func stream() -> AsyncStream<RealtimeEvent> { events }

    func start() {
        guard !running else { return }
        running = true
        Task { await runForever() }
    }

    func stop() {
        running = false
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    // MARK: - Sending

    /// Tells the account's other devices what this one is playing, or that it is
    /// idle. Sending this at all is what makes a device appear as a cast target.
    func publishPlayback(_ state: CastPlaybackState?) {
        var frame: [String: Any] = ["type": "playback"]
        // The key must be present even when null. A frame that simply omits it
        // still registers the receiver, but the same omission on a real state
        // would silently drop the position.
        if let state, let data = try? JSONEncoder().encode(state),
           let object = try? JSONSerialization.jsonObject(with: data) {
            frame["state"] = object
        } else {
            frame["state"] = NSNull()
        }
        send(frame)
    }

    func sendCommand(to sessionId: String, command: CastCommand) {
        send(["type": "command", "to": sessionId, "command": encode(command)])
    }

    private func encode(_ command: CastCommand) -> [String: Any] {
        switch command {
        case .pause: return ["action": "pause"]
        case .resume: return ["action": "resume"]
        case .stop: return ["action": "stop"]
        case .next: return ["action": "next"]
        case .seek(let positionMs): return ["action": "seek", "positionMs": positionMs]
        case .play(let request):
            guard let data = try? JSONEncoder().encode(request),
                  let object = try? JSONSerialization.jsonObject(with: data) else {
                return ["action": "play"]
            }
            return ["action": "play", "playback": object]
        }
    }

    private func send(_ frame: [String: Any]) {
        guard let task,
              let data = try? JSONSerialization.data(withJSONObject: frame),
              let text = String(data: data, encoding: .utf8) else { return }
        task.send(.string(text)) { _ in }
    }

    // MARK: - Connection

    private func runForever() async {
        var attempt = 0
        while running {
            let outcome = await runOnce()
            sessionId = nil

            if outcome.unauthorized { return }
            if outcome.tokenExpired {
                attempt = 0
                // Renew first. Reconnecting with the same dead token just loops.
                if await renew(outcome.token) == nil { return }
                continue
            }
            if outcome.sawReady { attempt = 0 }
            let delay = min(pow(2.0, Double(attempt)), 30)
            attempt += 1
            try? await Task.sleep(for: .seconds(delay))
        }
    }

    private struct Outcome {
        var sawReady = false
        var tokenExpired = false
        var unauthorized = false
        var token: String?
    }

    private func runOnce() async -> Outcome {
        guard let token = store.load()?.accessToken, !token.isEmpty else {
            return Outcome(unauthorized: true)
        }
        var outcome = Outcome(token: token)

        var request = URLRequest(url: endpoint)
        request.setValue(ApiConfig.clientKind, forHTTPHeaderField: ApiConfig.clientHeader)
        request.setValue(ApiConfig.userAgent, forHTTPHeaderField: "User-Agent")

        let task = URLSession.shared.webSocketTask(with: request)
        self.task = task
        task.resume()

        // Auth travels in the first frame rather than the query string, so
        // access tokens stay out of URLs and proxy access logs. The server
        // allows five seconds for it.
        send(["type": "auth", "token": token])

        while true {
            do {
                let message = try await task.receive()
                guard case .string(let text) = message,
                      let data = text.data(using: .utf8),
                      let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let type = object["type"] as? String else { continue }

                if type == "ready" {
                    outcome.sawReady = true
                    sessionId = object["sessionId"] as? String
                    // The handshake carries the receivers already connected, so
                    // a client that opens second still sees the television
                    // without waiting for it to report again.
                    if let list = receivers(from: object) {
                        continuation.yield(.receivers(list))
                    }
                    continue
                }
                if let event = decode(type: type, object: object) {
                    continuation.yield(event)
                }
            } catch {
                let code = task.closeCode.rawValue
                outcome.tokenExpired = (code == Self.closeTokenExpired)
                outcome.unauthorized = (code == Self.closeUnauthorized)
                self.task = nil
                return outcome
            }
        }
    }

    private func receivers(from object: [String: Any]) -> [CastReceiver]? {
        guard let raw = object["receivers"],
              let data = try? JSONSerialization.data(withJSONObject: raw),
              let list = try? JSONDecoder().decode([CastReceiver].self, from: data) else { return nil }
        return list
    }

    private func decode(type: String, object: [String: Any]) -> RealtimeEvent? {
        switch type {
        case "favorites":
            return .favorites(action: object["action"] as? String ?? "")
        case "progress":
            return .progress(
                action: object["action"] as? String ?? "",
                historyChanged: object["history"] as? Bool ?? false
            )
        case "receivers":
            return .receivers(receivers(from: object) ?? [])
        case "command":
            guard let raw = object["command"] as? [String: Any],
                  let command = decodeCommand(raw) else { return .unknown(type) }
            return .command(
                from: object["from"] as? String,
                fromName: object["fromName"] as? String,
                command: command
            )
        default:
            return .unknown(type)
        }
    }

    private func decodeCommand(_ raw: [String: Any]) -> CastCommand? {
        switch raw["action"] as? String {
        case "pause": return .pause
        case "resume": return .resume
        case "stop": return .stop
        case "next": return .next
        case "seek":
            guard let position = raw["positionMs"] as? Int else { return nil }
            return .seek(positionMs: position)
        case "play":
            guard let playback = raw["playback"],
                  let data = try? JSONSerialization.data(withJSONObject: playback),
                  let request = try? JSONDecoder().decode(CastPlayRequest.self, from: data) else { return nil }
            return .play(request)
        default:
            // An action this build does not know is ignored, not an error: a
            // newer server must not be able to drop an older client's socket.
            return nil
        }
    }
}

File: android/core/src/main/kotlin/com/streamhub/core/net/RealtimeClient.kt
package com.streamhub.core.net

import com.streamhub.core.model.CastCommand
import com.streamhub.core.model.CastPlaybackState
import com.streamhub.core.model.CastReceiver
import com.streamhub.core.model.Session
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.flow.channelFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

/**
 * Live library updates from `/api/realtime`.
 *
 * Collecting [events] opens the socket; cancelling the collection closes it. The
 * flow survives disconnects on its own, so a caller never has to reconnect.
 */
class RealtimeClient(
    baseUrl: String,
    private val store: SessionStore,
    private val renew: suspend (staleAccessToken: String?) -> Session?,
    private val client: OkHttpClient = OkHttpClient(),
    private val json: Json = Json { ignoreUnknownKeys = true },
    private val reconnectBaseMs: Long = 1_000,
    private val reconnectMaxMs: Long = 30_000,
) {

    private val endpoint: String = baseUrl.trimEnd('/').toHttpUrl().newBuilder()
        .addPathSegment("api")
        .addPathSegment("realtime")
        .build()
        .toString()

    private val _connected = MutableStateFlow(false)

    /**
     * Whether a socket is currently authenticated and carrying events. Useful in
     * a status view: live sync being down is a different problem from the server
     * being unreachable, and they should not look the same.
     */
    val connected: StateFlow<Boolean> = _connected.asStateFlow()

    private val _sessionId = MutableStateFlow<String?>(null)

    /**
     * This connection's own session id, as the server sees it.
     *
     * The receiver list includes every announced device on the account — this
     * one among them once it announces. Without knowing its own id a device
     * would offer itself as a target to cast to.
     */
    val sessionId: StateFlow<String?> = _sessionId.asStateFlow()

    /**
     * The socket currently carrying events, for the frames a client sends after
     * the handshake. Volatile because commands are sent from whichever thread
     * the UI is on, while this is assigned by the socket's own callback thread.
     */
    @Volatile
    private var live: WebSocket? = null

    /**
     * Tells the account's other devices what this one is playing, or that it is
     * idle. Sending this at all is what makes a device appear as a cast target,
     * so a television calls it once on startup with a null state.
     *
     * Returns false when there is no live socket — the caller decides whether
     * that is worth surfacing.
     */
    fun publishPlayback(state: CastPlaybackState?): Boolean {
        val socket = live ?: return false
        return socket.send(json.encodeToString(PlaybackFrame(type = "playback", state = state)))
    }

    /** Sends a command to one of the account's other devices. */
    fun sendCommand(to: String, command: CastCommand): Boolean {
        val socket = live ?: return false
        return socket.send(json.encodeToString(CommandFrame(type = "command", to = to, command = command)))
    }

    fun events(): Flow<RealtimeEvent> = channelFlow {
        var attempts = 0

        while (true) {
            val outcome = runSocket { trySend(it) }
            _connected.value = false

            when {
                // Nothing to authenticate with, or the server rejected the token
                // outright. Reconnecting cannot help; the app has to sign in.
                outcome.noSession || outcome.closeCode == CLOSE_UNAUTHORIZED -> return@channelFlow

                // The socket is only authenticated at the handshake, so the
                // server drops it the moment the token lapses. Renew *first* —
                // reconnecting with the same dead token just loops.
                outcome.closeCode == CLOSE_TOKEN_EXPIRED -> {
                    attempts = 0
                    if (renew(outcome.token) == null) return@channelFlow
                }

                else -> {
                    // A connection that reached "ready" was healthy, so treat a
                    // later drop as the first failure rather than staying at
                    // whatever backoff got us connected.
                    if (outcome.sawReady) attempts = 0
                    delay(backoffMs(attempts))
                    attempts += 1
                }
            }
        }

        awaitClose { }
    }.buffer(BUFFER)

    private fun backoffMs(attempts: Int): Long {
        val scaled = reconnectBaseMs shl attempts.coerceAtMost(16)
        return scaled.coerceIn(reconnectBaseMs, reconnectMaxMs)
    }

    /** Runs one connection and suspends until it closes. */
    private suspend fun runSocket(onEvent: (RealtimeEvent) -> Unit): Outcome {
        val token = store.load()?.accessToken?.takeIf { it.isNotEmpty() }
            ?: return Outcome(closeCode = null, sawReady = false, token = null, noSession = true)

        val closed = CompletableDeferred<Int?>()
        var sawReady = false

        val listener = object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                live = webSocket
                // Auth travels in the first frame rather than the query string,
                // so access tokens stay out of proxy logs. The server allows
                // five seconds for it.
                webSocket.send(json.encodeToString(AuthFrame(type = "auth", token = token)))
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                val frame = runCatching { json.decodeFromString<JsonObject>(text) }.getOrNull() ?: return
                val type = frame["type"]?.jsonPrimitive?.contentOrNull ?: return
                if (type == READY) {
                    sawReady = true
                    _connected.value = true
                    _sessionId.value = frame["sessionId"]?.jsonPrimitive?.contentOrNull
                    // The handshake carries the receivers already connected, so
                    // a phone that opens second still sees the television
                    // without waiting for it to report again.
                    receiversOf(frame)?.let { onEvent(RealtimeEvent.Receivers(it)) }
                    return
                }
                onEvent(toEvent(type, frame))
            }

            /** The server's code arrives here; onClosed would lose it. */
            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                closed.complete(code)
                webSocket.close(NORMAL_CLOSURE, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                closed.complete(code)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                closed.complete(null)
            }
        }

        val socket = client.newWebSocket(Request.Builder().url(endpoint).build(), listener)
        return try {
            Outcome(closeCode = closed.await(), sawReady = sawReady, token = token, noSession = false)
        } finally {
            socket.cancel()
            live = null
            _sessionId.value = null
        }
    }

    private fun receiversOf(frame: JsonObject): List<CastReceiver>? {
        val array = frame["receivers"] ?: return null
        return runCatching { json.decodeFromJsonElement<List<CastReceiver>>(array) }.getOrNull()
    }

    private fun toEvent(type: String, frame: JsonObject): RealtimeEvent {
        val action = frame["action"]?.jsonPrimitive?.contentOrNull.orEmpty()
        return when (type) {
            "favorites" -> RealtimeEvent.Favorites(action, frame["id"]?.jsonPrimitive?.contentOrNull)
            "progress" -> RealtimeEvent.Progress(
                action = action,
                historyChanged = frame["history"]?.jsonPrimitive?.booleanOrNull ?: false,
            )
            "receivers" -> RealtimeEvent.Receivers(receiversOf(frame).orEmpty())
            "command" -> {
                // An action this build does not implement decodes to nothing;
                // an older client should ignore it, not drop the connection.
                val command = frame["command"]
                    ?.let { runCatching { json.decodeFromJsonElement<CastCommand>(it) }.getOrNull() }
                if (command == null) RealtimeEvent.Unknown(type)
                else RealtimeEvent.Command(
                    from = frame["from"]?.jsonPrimitive?.contentOrNull,
                    fromName = frame["fromName"]?.jsonPrimitive?.contentOrNull,
                    command = command,
                )
            }
            else -> RealtimeEvent.Unknown(type)
        }
    }

    private class Outcome(
        val closeCode: Int?,
        val sawReady: Boolean,
        val token: String?,
        val noSession: Boolean,
    )

    /**
     * `type` deliberately has no default value. kotlinx.serialization omits
     * properties that equal their default, so a default here would send
     * `{"token":"…"}` with no type, and the server closes such a frame as
     * unauthorized — the socket would never connect at all.
     */
    @Serializable
    private data class AuthFrame(val type: String, val token: String)

    /**
     * No defaults here either, and for the same reason: a `state` that defaulted
     * to null would be omitted from the frame entirely.
     */
    @Serializable
    private data class PlaybackFrame(val type: String, val state: CastPlaybackState?)

    @Serializable
    private data class CommandFrame(val type: String, val to: String, val command: CastCommand)

    private companion object {
        const val READY = "ready"
        const val NORMAL_CLOSURE = 1000
        const val BUFFER = 64

        /** The server closes with this once the token behind the handshake lapses. */
        const val CLOSE_TOKEN_EXPIRED = 4002
        const val CLOSE_UNAUTHORIZED = 4003
    }
}
